src/editor/Model/FileExplorer.re

```diff
@@ -0,0 +1,186 @@
+open Revery;
+open Oni_Core;
+
+type t = {
+  directory: UiTree.t,
+  isOpen: bool,
+};
+
+module ExplorerId =
+  UniqueId.Make({});
+
+let toFsNode = node =>
+  switch (node) {
+  | UiTree.FileSystemNode(n) => n
+  };
+
+let getFileIcon = (languageInfo, iconTheme, filePath) => {
+  let fileIcon =
+    LanguageInfo.getLanguageFromFilePath(languageInfo, filePath)
+    |> IconTheme.getIconForFile(iconTheme, filePath);
+
+  switch (fileIcon) {
+  | Some(_) as x => x
+  | None => None
+  };
+};
+
+let createFsNode =
+    (~children, ~depth, ~path, ~displayName, ~fileIcon, ~isDirectory) => {
+  /**
+     TODO: Find an icon theme with folders and use those icons
+     Fallbacks are used for the directory icons. FontAwesome is not
+     accessible here so we specify no icons for directories.
+   */
+  let (primary, secondary) = isDirectory ? (None, None) : (fileIcon, None);
+
+  UiTree.FileSystemNode({
+    path,
+    depth,
+    displayName,
+    children,
+    isDirectory,
+    icon: primary,
+    secondaryIcon: secondary,
+  });
+};
+
+let isDir = path => Sys.file_exists(path) ? Sys.is_directory(path) : false;
+
+/**
+  getFilesAndFolders
+
+   This function uses Lwt to get all the files and folders in a directory
+   then for each we check if it is a file or folder.
+   if it is a directory we recursively call getFilesAndFolders on it
+   to resolves its subfolders and files. We do this concurrently using
+   Lwt_list.map_p. The recursion is gated by the depth value so it does
+   not recurse too far.
+ */
+let getFilesAndFolders = (~maxDepth, ~ignored, cwd, getIcon) => {
+  let rec getDirContent = (~depth, cwd, getIcon) => {
+    Lwt_unix.files_of_directory(cwd)
+    /* Filter out the relative name for current and parent directory*/
+    |> Lwt_stream.filter(name => name != ".." && name != ".")
+    /* Remove ignored files from search */
+    |> Lwt_stream.filter(name => !List.mem(name, ignored))
+    |> Lwt_stream.to_list
+    |> (
+      promise =>
+        Lwt.bind(promise, files =>
+          Lwt_list.map_p(
+            file => {
+              let path = Filename.concat(cwd, file);
+              let isDirectory = isDir(path);
+              let nextDepth = depth + 1;
+              let notMaximumDepth = nextDepth < maxDepth;
+
+              let%lwt children =
+                isDirectory && notMaximumDepth
```

Comment:
I've added the depth concept to this function so we can control how far this function recurses.

If run from a very deep directory we hit `EMFILE` error which is a unix error that says that our process has too many files open, this limit I think can be increased and I've seen a similar issue on vscode on my system.

With this though we only got 10 levels deep and if a user is trying to go further then when they expand a node at depth level 10, we fetch its contents (pretty seamlessly perceived performance wise. I didn't actually believe it was working so tested it a bunch, its super fast and imperceptible)